clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -508,17 +511,16 @@ private static int parseAcks(String acksString) {
      * Wait for cluster metadata including partitions for the given topic to be available.
      * @param topic The topic we want metadata for
      * @param maxWaitMs The maximum time in ms for waiting on the metadata
-     * @return The amount of time we waited in ms
+     * @param beginTimeMs Start time from which maxWaitMs applies
+     * @return The cluster containing topic metadata
      */
-    private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
+    private Cluster waitOnMetadata(String topic, long maxWaitMs, long beginTimeMs) throws InterruptedException {
         // add topic to metadata topic list if it is not there already and reset expiry
         this.metadata.add(topic);
-        if (metadata.fetch().partitionsForTopic(topic) != null)
-            return 0;
```

Comment:
Thanks. I prefer to avoid using `Atomic*` classes when we don't need atomicity guarantees. I introduced a `LongRef` class in the broker for this reason.

I think the right solution long-term is probably to introduce a class to complement `Time` that caches the last returned value so that one can choose when to refresh the value (for performance reasons). This was suggested some time ago in a different PR.

For this PR, as per your suggestion, I think using a private static class with two public fields is probably the way forward (since Java doesn't have tuples).